# PreVEAB
A computational package targets to measure the strength of passage adaptation and predict efficacy of a candidate vaccine strain based on its nucleotide sequence


## Background
Vaccine efficacy for the H3N2 Influenza virus has been alarmingly decreasing in recent years, but underlying causes for this observation are still elusive. Passage adaptation refers to substitutions accumulated during vaccine production might contribute to the vaccine efficacy. Using a statistical method called mutational mapping and 32,278 H3N2 hemagglutinin sequences, we identified 14 important codons driving passage adaptation in the embryonated egg. Based on enrichment scores derived from these codons, we defined an adaptive distance (AD) calibrating the strength of positive selection on each influenza strain. A very strong negative relationship between adaptive distance and vaccine efficacy (R-square=0.776) indicate the essential impact of passage adaptation on vaccine efficacy. For the first time, we provide a direct and comprehensive evidence that passage adaptation in the embryonated egg is a major determinant of the recent reduction in vaccine efficacy.


## Installing

PreVEAB is written in perl and R.

PreVEAB can be downloaded from github: https://github.com/Emma-CH/PreVEAB


## Running PreVEAB

To run the program, a coalescent tree which captures the ancestral relationships
of the tumor cells in a sample is required. The input tree file should be in the
newick format, which can be generated by the [ms
program](http://home.uchicago.edu/rhudson1/source/mksamples.html) with the `-T`
option. ms program has the full-assemblage of the options needed to generate
complex demographic histories of the sample. 


##### Notes
By default, the branch length from the ms program is measured in 4N
generations. CSiTE will simulate SNVs/CNVs events following the Poisson process.
For example, if we set -r 100, this is equivalent to set the population rescaled
parameter θ=4Nu as 100 (see ms manual for details). 

### Input files

#### Tree file (-t/--tree)

Tree file should be in newick format, for example:

    ((2:0.083,4:0.083):0.345,(5:0.322,(1:0.030,3:0.030):0.292):0.105);

#### Trunk variants file (--trunk\_vars)

You can simulate truncal variants by specify `--trunk_length`, or import trunk
variants directly using `--trunk_vars` option.  The file format of truncal
variants is like:
    
    #parental start end copy
    0 1 1 0
    1 12 123 +3 
    0 12 123 -1

- **parental**:  which parental copy the variant locates in (0 means one of the
  parental copy, 1 means the other copy)
- **start**:     the start position of the variant
- **end**:       the end position of the variant
- **copy**:      an integer. 0: SNV, -1: deletion, +int: amplification

P.S. start and end are 0 based. And the region of each variant is similar to the
bed file (inclusive on the left, but exclusive on the right end,
i.e.\[start,end)).

### Output files

#### SNVs file (-S/--snv)

This file contains the frequency information of simulated SNVs. It contains four
columns. 
- **position**:        the position (0-based) of SNVs
- **true\_freq**:      the true frequency of the alternative allele in the
  sample. 
- **total\_depth**:    the simulated total coverage of tumor and normal cells at
  the position of SNV. 
- **simulated\_freq**: the observed frequency of alternative allele across tumor
  and normal cell population.

#### Genotype file (--snv\_genotype)

This file contains the snv\_genotype of each tumor cell at SNV loci. Each SNV
has one record. The first column is the coordinate of the SNV. Subsequently,
there is one column for each tumor cell. The snv\_genotype is in the form of
‘M:N’. M denotes the number of alternative allele and N denotes the number of
reference allele.

### Examples

* Simulate the coalescent tree of a sample of 1000 tumor cells, which are
sampled from a exponetially growing tumor. 
(consult the manual of ms for more information)

    `ms 1000 1 -T -G 1 |tail -n1 > ms_tree.txt`

* Simulate the somatic SNVs of this sample. We assume the sequencing depth is
60, and the purity of the sample is 0.8, which means there are 250 normal
cells other than these 1000 tumor cells. Other settings are: 
a) the mutation rate of SNVs and CNVs are 10 and 0.1 respectively; 
b) the sequence length is 135534747 (chr10 of hg19); 
c) the cells of the sample are diploid. We save the
frequncy of the simulated SNVs to file 'snvs\_freq.txt'.

    `csite.py -t ms_tree.txt -P 0.8 --length 135534747 -r 10 -R 0.1 -D 60 -S
    snvs_freq.txt`

* There are no truncal muations in the simulation above. If you want to simulate
truncal muations, use the option `--trunk_length`.

    `csite.py -t ms_tree.txt -P 0.8 --length 135534747 -r 10 -R 0.1 -D 60 -S
    snvs_freq.txt --trunk_length 2.0`

* If you want to ignore the variants with the frequency <=0.01, you can use
`--prune 20` or `--prune_proportion 0.02` (we use `--prune 20` instead of
`--prune 10` for the cells are diploid in the our simulation). These two
options can be used to accelerate the simulation when your tree is huge.

    `csite.py -t ms_tree.txt -P 0.8 --length 135534747 -r 10 -R 0.1 -D 60 -S
    snvs_freq.txt --trunk_length 2.0 --prune 20`


## Author

* Hui Chen : chenh1@gis.a-star.edu.sg

## Corresponding author

*   Weiwei Zhai : zhaiww11@gis.a-star.edu.sg

## License

This project is licensed under the GNU GPLv3 License - see the
[LICENSE](LICENSE) file for details.
